google-cloud-spanner/src/main/java/com/google/cloud/spanner/util/DirectedReadsUtil.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spanner.util;
+
+import com.google.api.core.InternalApi;
+import com.google.cloud.spanner.ErrorCode;
+import com.google.cloud.spanner.SpannerExceptionFactory;
+import com.google.spanner.v1.DirectedReadOptions;
+
+/** Utility methods for DirectedReads feature. */
+@InternalApi
+public class DirectedReadsUtil {
+  static final int MAX_REPLICA_SELECTIONS_COUNT = 10;
+
+  public static DirectedReadOptions validateDirectedReadOptions(
+      DirectedReadOptions directedReadOptions) {
+    if (directedReadOptions.hasIncludeReplicas() && directedReadOptions.hasExcludeReplicas()) {
+      throw SpannerExceptionFactory.newSpannerException(
+          ErrorCode.INVALID_ARGUMENT,
+          "Only one of include_replicas or exclude_replicas can be set");
+    }
+    if ((directedReadOptions.hasIncludeReplicas()
+            && directedReadOptions.getIncludeReplicas().getReplicaSelectionsCount()
+                > MAX_REPLICA_SELECTIONS_COUNT)
+        || (directedReadOptions.hasExcludeReplicas()
+            && directedReadOptions.getExcludeReplicas().getReplicaSelectionsCount()
+                > MAX_REPLICA_SELECTIONS_COUNT)) {
+      throw SpannerExceptionFactory.newSpannerException(
+          ErrorCode.INVALID_ARGUMENT,
+          String.format(
+              "Maximum length of replica selection allowed in IncludeReplicas/ExcludeReplicas is %d",
+              MAX_REPLICA_SELECTIONS_COUNT));
+    }
+    return directedReadOptions;
+  }
+
+  public static DirectedReadOptions validateAndGetPreferredDirectedReadOptions(
+      DirectedReadOptions directedReadOptionsForClient,
+      DirectedReadOptions directedReadOptionsForRequest,
+      boolean readOnly) {
+    if (!readOnly) {
```

Comment:
This will end up failing all the RW transactions if the directed read options is set for the client. IMO that shouldn't be the right behaviour. Instead this setting should be silently ignored for RW transactions and considered only for RO transactions.

@olavloite WDYT? 

If we avoid adding this validation (and avoid exception) apart from improving the customer experience, we will also simplify a lot of things done in this PR (for ex - introducing `readOnly` across multiple method interfaces) .

Reply:
I don't remember exactly how this feature works, but in general I would say:
1. If the configuration is set at the client level (so configure once for all transactions executed by this client): This check should be removed.
2. If the option is set at the request level, which for a read/write transaction would mean something like:
   1. Start read/write transaction
   2. Execute a query with a `DirectedReadOption` set.
   3. Then we should fail at this point instead of silently ignoring, as the user is explicitly trying to use a feature with a read/write transaction that is not supported in that case. Silently ignoring might cause the user to believe that the feature is working.

Reply:
The configuration can be set at both the client as well as request level. The flexibility is provided at client level so that it can be treated as a global option and avoid the need to set it per RPC. If custom sets the config both at client as well as request, then the option at client is ignored in favour of the option set at the request.